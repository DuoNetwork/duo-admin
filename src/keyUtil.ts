import Storage from '@google-cloud/storage';
import { Aws } from 'aws-cli-js';
import * as fs from 'fs';
import { IKey, IOption, ISqlAuth } from './types';
import util from './util';
//const Storage = require('@google-cloud/storage');

class KeyUtil {
	public async getAwsKey(name: string) {
		const aws = new Aws();
		return aws.command(
			'ssm get-parameter --name ' + name + ' --region ap-southeast-1 --with-decryption'
		);
	}

	public async getAzureKey(name: string) {
		const baseUrl =
			'http://169.254.169.254/metadata/identity/oauth2/token?api-version=2018-02-01&resource=https%3A%2F%2Fvault.azure.net';

		const response: any = await util.get(baseUrl, { Metadata: 'true' });
		const responseJson = JSON.parse(response);
		const savedAccessToken = 'Bearer ' + responseJson.access_token;
		const url =
			name === 'price-feed-private'
				? 'https://price-dev-test.vault.azure.net/secrets/price-feed-private?api-version=2016-10-01'
				: 'https://price-dev-test.vault.azure.net/secrets/MySQL-DB-Dev/653497bb15104ba8aa0eccfa4a90b612';
		const bodyKey: any = await util.get(url, { Authorization: savedAccessToken });

		const responseKeyJson = JSON.parse(bodyKey);
		// console.log('Private Key: ' + responseKeyJson.value);
		return responseKeyJson.value;
	}

	public async getGcpKey(name: string): Promise<any> {
		const storage = Storage({
			projectId: 'duo-network'
		});

		const bucketName = 'eth-test';
		const fileName = name === 'price-feed-private' ? 'testkey.txt' : 'MySQL_DB_Dev.txt';
		return new Promise((resolve, reject) => {
			storage
				.bucket(bucketName)
				.file(fileName)
				.download()
				.then(data => resolve(data.toString()))
				.catch(err => reject(err));
		});
	}
	public async getKey(option: IOption): Promise<IKey> {
		if (!option.live && !option.server) {
			const key = option.azure
				? JSON.parse(fs.readFileSync('./keys/kovan/pfAzure.json').toString())
				: option.gcp
					? JSON.parse(fs.readFileSync('./keys/kovan/pfGcp.json').toString())
					: JSON.parse(fs.readFileSync('./keys/kovan/pfAws.json').toString());
			return {
				publicKey: key.publicKey,
				privateKey: key.privateKey
			};
		} else {
			let key;
			if (option.aws) {
				const keyData = await this.getAwsKey('price-feed-private');
				key = JSON.parse(keyData.object.Parameter.Value);
			}
			if (option.azure) {
				const keyData = await this.getAzureKey('price-feed-private');
				key = JSON.parse(keyData);
			}
			if (option.gcp) {
				const keyData = await this.getGcpKey('price-feed-private');
				key = JSON.parse(keyData);
			}
			return {
				publicKey: key['publicKey'],
				privateKey: key['privateKey']
			};
		}
	}

	public async getSqlAuth(option: IOption): Promise<ISqlAuth> {
		let key;
		if (!option.live && !option.server) {
			const mysqlAuthFile = JSON.parse(fs.readFileSync('./keys/mysql.json').toString());
			return {
				host: mysqlAuthFile.host,
				user: mysqlAuthFile.user,
				password: mysqlAuthFile.password
			};
		} else {
			if (option.aws) {
				const keyData = await this.getAwsKey('MySQL_DB_Dev');
				key = JSON.parse(keyData.object.Parameter.Value);
			} else if (option.azure) {
				const keyData = await this.getAzureKey('MySQL_DB_Dev');
				key = JSON.parse(keyData);
			} else {
				const keyData = await this.getGcpKey('MySQL_DB_Dev');
				key = JSON.parse(keyData);
			}
			return {
				host: key['host'],
				user: key['user'],
				password: key['password']
			};
		}
	}
}

const keyUtil = new KeyUtil();
export default keyUtil;
