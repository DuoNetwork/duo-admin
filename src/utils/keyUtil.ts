import { Storage } from '@google-cloud/storage';
import { Aws } from 'aws-cli-js';
import * as CST from '../common/constants';
import { IKey, IOption, ISqlAuth } from '../common/types';
import httpUtil from './httpUtil';
import util from './util';

class KeyUtil {
	public async getAwsKey(name: string) {
		const aws = new Aws();
		const keyData = await aws.command(
			'ssm get-parameter --name ' + name + ' --region ap-southeast-1 --with-decryption'
		);

		return keyData.object.Parameter.Value;
	}

	public async getAzureKey(name: string, live: boolean): Promise<string> {
		const baseUrl = CST.KEY_AZURE_BASE_URL;

		const response: any = await httpUtil.get(baseUrl, { Metadata: 'true' });
		const responseJson = JSON.parse(response);
		const savedAccessToken = 'Bearer ' + responseJson.access_token;
		const url = live
			? `${CST.KEY_AZURE_URL_LIVE}${name}${CST.KEY_AZURE_API_VERSION}`
			: `${CST.KEY_AZURE_URL_DEV}${name}${CST.KEY_AZURE_API_VERSION}`;
		const bodyKey: any = await httpUtil.get(url, { Authorization: savedAccessToken });

		const responseKeyJson = JSON.parse(bodyKey);
		return responseKeyJson.value;
	}

	public async getGcpKey(name: string, live: boolean): Promise<string> {
		const storage = new Storage({
			projectId: live ? CST.KEY_GCP_PROJECT_ID_LIVE : CST.KEY_GCP_PROJECT_ID_DEV
		});

		const bucketName = live ? CST.KEY_GCP_BUKKET_NAME_LIVE : CST.KEY_GCP_BUKKET_NAME_DEV;
		const fileName = name + '.txt';
		return (storage
			.bucket(bucketName)
			.file(fileName)
			.download() as any).then((data: any) => data.toString());
	}

	public async getKey(tool: string, option: IOption): Promise<IKey> {
		let key = {
			publicKey: '',
			privateKey: ''
		};
		if (!option.live && !option.server) {
			try {
				key = option.azure
					? require('../keys/kovan/pfAzure.json')
					: option.gcp
					? require('../keys/kovan/pfGcp.json')
					: require('../keys/kovan/pfAws.json');
			} catch (error) {
				util.logError(error);
			}
			return {
				publicKey: key.publicKey,
				privateKey: key.privateKey
			};
		} else {
			if (option.aws)
				key = JSON.parse(
					await this.getAwsKey(
						option.live ? CST.KEY_ETH_NAME_LIVE : CST.KEY_ETH_NAME_DEV
					)
				)[tool];
			else if (option.azure)
				key = JSON.parse(
					await this.getAzureKey(
						option.live ? CST.KEY_ETH_NAME_LIVE : CST.KEY_ETH_NAME_DEV,
						option.live
					)
				)[tool];
			else if (option.gcp)
				key = JSON.parse(
					await this.getGcpKey(
						option.live ? CST.KEY_ETH_NAME_LIVE : CST.KEY_ETH_NAME_DEV,
						option.live
					)
				)[tool];

			return {
				publicKey: key.publicKey,
				privateKey: key.privateKey
			};
		}
	}

	public async getSqlAuth(option: IOption): Promise<ISqlAuth> {
		let key = {
			host: '',
			user: '',
			password: ''
		};
		if (!option.live && !option.server) {
			try {
				key = require('../keys/mysql.json');
			} catch (error) {
				util.logError(error);
			}
			return {
				host: key.host,
				user: key.user,
				password: key.password
			};
		} else {
			if (option.aws)
				key = JSON.parse(
					await this.getAwsKey(
						option.live ? CST.KEY_SQL_NAME_LIVE : CST.KEY_SQL_NAME_DEV
					)
				);
			else if (option.azure)
				key = JSON.parse(
					await this.getAzureKey(
						option.live ? CST.KEY_SQL_NAME_LIVE : CST.KEY_SQL_NAME_DEV,
						option.live
					)
				);
			else if (option.gcp)
				key = JSON.parse(
					await this.getGcpKey(
						option.live ? CST.KEY_SQL_NAME_LIVE : CST.KEY_SQL_NAME_DEV,
						option.live
					)
				);

			return {
				host: key['host'],
				user: key['user'],
				password: key['password']
			};
		}
	}
}

const keyUtil = new KeyUtil();
export default keyUtil;
