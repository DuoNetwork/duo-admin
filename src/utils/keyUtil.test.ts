import keyUtil from './keyUtil';

test('getKey server and live', async () => {
	const keyString = JSON.stringify({
		publicKey: 'publicKey',
		privateKey: 'privateKey'
	});
	keyUtil.getAwsKey = jest.fn(() => Promise.resolve(keyString));
	keyUtil.getAzureKey = jest.fn(() => Promise.resolve(keyString));
	keyUtil.getGcpKey = jest.fn(() => Promise.resolve(keyString));
	await keyUtil.getKey({
		live: true,
		server: true,
		aws: true
	} as any);
	await keyUtil.getKey({
		live: true,
		server: true,
		azure: true
	} as any);
	await keyUtil.getKey({
		live: true,
		server: true,
		gcp: true
	} as any);
	await keyUtil.getKey({
		live: true,
		server: true
	} as any);
	expect((keyUtil.getAwsKey as jest.Mock).mock.calls).toMatchSnapshot();
	expect((keyUtil.getAzureKey as jest.Mock).mock.calls).toMatchSnapshot();
	expect((keyUtil.getGcpKey as jest.Mock).mock.calls).toMatchSnapshot();
});

test('getKey dev', async () => {
	const keyString = JSON.stringify({
		publicKey: 'publicKey',
		privateKey: 'privateKey'
	});
	keyUtil.getAwsKey = jest.fn(() => Promise.resolve(keyString));
	keyUtil.getAzureKey = jest.fn(() => Promise.resolve(keyString));
	keyUtil.getGcpKey = jest.fn(() => Promise.resolve(keyString));
	await keyUtil.getKey({
		aws: true
	} as any);
	await keyUtil.getKey({
		azure: true
	} as any);
	await keyUtil.getKey({
		gcp: true
	} as any);
	expect(keyUtil.getAwsKey as jest.Mock).not.toBeCalled();
	expect(keyUtil.getAzureKey as jest.Mock).not.toBeCalled();
	expect(keyUtil.getGcpKey as jest.Mock).not.toBeCalled();
});

test('getSqlAuth server and live', async () => {
	const keyString = JSON.stringify({
		host: 'host',
		user: 'user',
		password: 'password'
	});
	keyUtil.getAwsKey = jest.fn(() => Promise.resolve(keyString));
	keyUtil.getAzureKey = jest.fn(() => Promise.resolve(keyString));
	keyUtil.getGcpKey = jest.fn(() => Promise.resolve(keyString));
	await keyUtil.getSqlAuth({
		live: true,
		server: true,
		aws: true
	} as any);
	await keyUtil.getSqlAuth({
		live: true,
		server: true,
		azure: true
	} as any);
	await keyUtil.getSqlAuth({
		live: true,
		server: true,
		gcp: true
	} as any);
	await keyUtil.getSqlAuth({
		live: true,
		server: true
	} as any);
	expect((keyUtil.getAwsKey as jest.Mock).mock.calls).toMatchSnapshot();
	expect((keyUtil.getAzureKey as jest.Mock).mock.calls).toMatchSnapshot();
	expect((keyUtil.getGcpKey as jest.Mock).mock.calls).toMatchSnapshot();
});

test('getSqlAuth dev', async () => {
	const keyString = JSON.stringify({
		host: 'host',
		user: 'user',
		password: 'password'
	});
	keyUtil.getAwsKey = jest.fn(() => Promise.resolve(keyString));
	keyUtil.getAzureKey = jest.fn(() => Promise.resolve(keyString));
	keyUtil.getGcpKey = jest.fn(() => Promise.resolve(keyString));
	await keyUtil.getSqlAuth({
		aws: true
	} as any);
	await keyUtil.getSqlAuth({
		azure: true
	} as any);
	await keyUtil.getSqlAuth({
		gcp: true
	} as any);
	expect(keyUtil.getAwsKey as jest.Mock).not.toBeCalled();
	expect(keyUtil.getAzureKey as jest.Mock).not.toBeCalled();
	expect(keyUtil.getGcpKey as jest.Mock).not.toBeCalled();
});
