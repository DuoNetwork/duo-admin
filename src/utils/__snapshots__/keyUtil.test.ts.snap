// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getKey server and live 1`] = `
Array [
  Array [
    "price-feed-private-Live",
  ],
]
`;

exports[`getKey server and live 2`] = `
Array [
  Array [
    "price-feed-private-Live",
    true,
  ],
]
`;

exports[`getKey server and live 3`] = `
Array [
  Array [
    "price-feed-private-Live",
    true,
  ],
]
`;

exports[`getSqlAuth server and live 1`] = `
Array [
  Array [
    "MySQL-DB-Live",
  ],
]
`;

exports[`getSqlAuth server and live 2`] = `
Array [
  Array [
    "MySQL-DB-Live",
    true,
  ],
]
`;

exports[`getSqlAuth server and live 3`] = `
Array [
  Array [
    "MySQL-DB-Live",
    true,
  ],
]
`;
